database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala

```diff
@@ -123,20 +111,79 @@ trait WorkflowMetadataSummaryEntryComponent {
     }
   }
 
-  /**
-    * Filters workflows associated with the labels, combining the label matches using using op.
-    *
-    * The op can OR together or AND together the filtered rows.
-    */
-  private def existsWorkflowLabels(workflowMetadataSummaryEntry: WorkflowMetadataSummaryEntries,
-                                   labelKeyValues: Set[(String, String)],
-                                   op: (Rep[Boolean], Rep[Boolean]) => Rep[Boolean]): Option[Rep[Boolean]] = {
-    labelKeyValues
-      .map({
-        case (labelKey, labelValue) =>
-          existsWorkflowIdLabelKeyAndValue(workflowMetadataSummaryEntry.workflowExecutionUuid, labelKey, labelValue)
-      })
-      .reduceLeftOption(op)
+  private def doFiltering(parentIdWorkflowMetadataKey: String,
+                          workflowStatuses: Set[String], workflowNames: Set[String],
+                          workflowExecutionUuids: Set[String],
+                          labelAndKeyLabelValues: Set[(String,String)],
+                          labelOrKeyLabelValues: Set[(String,String)],
+                          excludeLabelAndValues: Set[(String,String)],
+                          excludeLabelOrValues: Set[(String,String)],
+                          submissionTimestampOption: Option[Timestamp],
+                          startTimestampOption: Option[Timestamp],
+                          endTimestampOption: Option[Timestamp],
+                          includeSubworkflows: Boolean) = {
+
+    val filter = filterWorkflowMetadataSummaryEntries(
+      parentIdWorkflowMetadataKey,
+      workflowStatuses,
+      workflowNames,
+      workflowExecutionUuids,
+      submissionTimestampOption,
+      startTimestampOption,
+      endTimestampOption,
+      includeSubworkflows
+    )
+
+    // First do the summary table filtering, then some performance optimizations below for labels.
+    val baseQuery = workflowMetadataSummaryEntries.filter(filter)
+
+    // Label ANDs and ORs can be expressed as `join`s which at least on MySQL appear to execute more efficiently than
+    // equivalent `exists` formulations.
+    type LabelsQuery = Query[CustomLabelEntries, CustomLabelEntry, Seq]
```

Comment:
Why `Seq` here and `Set` below?

Reply:
The label kv inputs are `Set`s and the `Query`s collections parallel those. The `Seq`s come from the Slick API.